### PR TITLE
Support Autotune

### PR DESCRIPTION
**Pull request by Pzzzzz5142**
~~This pull request will cause the calculation to regress. It will remain a work in progress until I resolve the calculation issue.~~

~~Calculation issue is fixed. Now autotune will cause memory buffer out of order. Working on this issue.~~

Ready to merge.

### CODE REVIEW

**Review by umiswing — CHANGES_REQUESTED** [review]

File: lib/ExecutionEngine/HandsOnNVGPURunnerUtils.cu [hunk 1]
```diff
@@ -149,13 +208,70 @@ void cutlassGemmWithVarMeanF16(int64_t rankA, void *dstA, int64_t rankB,
...
+  using namespace mlir::hands_on_mlir::homnvgpu_kernel;
+  GemmNVTERunner<half> gemm;
+
+  checkCudaErrors(cudaStreamSynchronize(nullptr));
 
   auto status = gemm.run(rankA, dstA, transa, rankB, dstB, transb, rankC, dstC,
                          rankD, dstD, activation, alpha, beta);
 
-  assert(status == cutlass::Status::kSuccess);
+  return;
```
Comment (Pzzzzz5142) [inline lib/ExecutionEngine/HandsOnNVGPURunnerUtils.cu:220]: remove the debug code

**Comment by Pzzzzz5142** [comment]
The calculation issue should be fixed in #5.

File: include/NVGPUKernels/LayernormGemmRunner.h [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+                                      rankD, dstD, rankVar, dstVar, rankMean,
+                                      dstMean, alpha, beta, eps, args);
+    if (status != Status::kSuccess) {
+      return status;
+    }
+
+    status = initialize_and_update(args);
+    if (status != Status::kSuccess) {
+      return status;
+    }
```
[the remark this answers is not in the document]
Comment (Pzzzzz5142) [reply include/NVGPUKernels/LayernormGemmRunner.h]: Unrelated to this pr.

File: include/NVGPUKernels/LayernormGemmRunner.h [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+                     {ElementC(alpha), ElementD(beta)}, eps);
+
+    params_ = Params(args);
+
+    return Status::kSuccess;
+  }
+
+  Status initialize_and_update(const Arguments &args) {
+    return Status::kSuccess;
+  }
```
[the remark this answers is not in the document]
Comment (Pzzzzz5142) [reply include/NVGPUKernels/LayernormGemmRunner.h]: Fixed

File: include/NVGPUKernels/LayernormGemmRunner.h [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+      }
+      auto error = cudaMalloc(&device_workspace_, required_workspace);
+      assert(error == cudaSuccess);
+      device_workspace_size_ = required_workspace;
+    }
+
+    if (first_initialize) {
+      return op->initialize(args, device_workspace_);
+    } else {
+      return op->update(args);
```
[the remark this answers is not in the document]
Comment (Pzzzzz5142) [reply include/NVGPUKernels/LayernormGemmRunner.h]: Fixed

**Review by umiswing — APPROVED** [review]

**Pull request merged**